fvm/environment/event_emitter.go

```diff
@@ -115,37 +112,38 @@ func (emitter *eventEmitter) EmitEvent(event cadence.Event) error {
 
 	payloadSize := uint64(len(payload))
 
-	// skip limit if payer is service account
-	if emitter.payer != emitter.chain.ServiceAddress() {
-		totalSize := emitter.eventCollection.TotalByteSize() + payloadSize
-		if totalSize > emitter.eventCollectionByteSizeLimit {
-			return errors.NewEventLimitExceededError(
-				totalSize,
-				emitter.eventCollectionByteSizeLimit)
-		}
-	}
-
 	flowEvent := flow.Event{
 		Type:             flow.EventType(event.EventType.ID()),
 		TransactionID:    emitter.txID,
 		TransactionIndex: emitter.txIndex,
-		EventIndex:       emitter.eventCollection.eventCounter,
+		EventIndex:       emitter.eventCollection.TotalEventCounter(),
 		Payload:          payload,
 	}
 
+	isServiceAccount := emitter.payer == emitter.chain.ServiceAddress()
+
 	if emitter.serviceEventCollectionEnabled {
 		ok, err := IsServiceEvent(event, emitter.chain.ChainID())
 		if err != nil {
 			return fmt.Errorf("unable to check service event: %w", err)
 		}
 		if ok {
-			emitter.eventCollection.AppendServiceEvent(flowEvent, payloadSize)
+			eventEmitError := emitter.eventCollection.AppendServiceEvent(flowEvent, payloadSize)
+			// skip limit if payer is service account
+			if !isServiceAccount && eventEmitError != nil {
+				return eventEmitError
+			}
 		}
 		// We don't return and append the service event into event collection
 		// as well.
 	}
 
-	emitter.eventCollection.AppendEvent(flowEvent, payloadSize)
+	eventEmitError := emitter.eventCollection.AppendEvent(flowEvent, payloadSize)
+	// skip limit if payer is service account
+	if !isServiceAccount && eventEmitError != nil {
```

Comment:
In a future PR we should consider getting rid of this `isServiceAccount` check here, and instead set the event limit to maxuint if this is the service account.

Reply:
+1.  we can extend ShouldDisableMemoryAndInteractionLimits to ShouldDisableNonComputeLimit or something